test: Embed univalue json tests in binary

maflcko opened the pull request:
All other benchmarks and tests have their data embedded, except for the univalue json tests.

This is not only confusing, but also problematic, when the test binary is moved to a different system for testing, because one has to put the test files in the source dir that was used at compile-time.

Fix all issues by embedding them. Also, re-enable a disabled test. Also, fix an issue in the GenerateHeaderFromJson.cmake.

Requested in https://github.com/bitcoin/bitcoin/pull/31434/files#r1876000910

Commit:
test: Re-enable univalue test fail18.json

Also, extend the pass2.json test to the maximum depth possible. The two
tests are now similar to fail45.json and pass4.json, except for the
string element in the inner-most array.

Also, sort.

Commit:
test: Embed univalue json tests in binary

TheCharlatan commented:
Nice, Concept ACK

hebasto commented:
Concept ACK. Thank you for picking this up!

sedited reviewed (approved):
ACK fabda022536c7a000a575bbb05059d27d29b5cca

hebasto commented on src/univalue/CMakeLists.txt:
```diff
@@ -15,11 +15,120 @@ target_include_directories(univalue
-      JSON_TEST_SRC=\"${CMAKE_CURRENT_SOURCE_DIR}/test\"
-  )
+  include(GenerateHeaders)
+  generate_header_from_json(data/base58_encode_decode.json)
```
Drop this one?

hebasto commented on src/univalue/CMakeLists.txt:
```diff
@@ -15,11 +15,120 @@ target_include_directories(univalue
+  generate_header_from_json(test/fail42.json)
+  generate_header_from_json(test/fail44.json)
+  generate_header_from_json(test/fail45.json)
+  generate_header_from_json(test/fail3.json)
```
nit: Reorder?

hebasto commented on src/univalue/CMakeLists.txt:
```diff
@@ -15,11 +15,120 @@ target_include_directories(univalue
+    ${CMAKE_CURRENT_BINARY_DIR}/test/round5.json.h
+    ${CMAKE_CURRENT_BINARY_DIR}/test/round6.json.h
+    ${CMAKE_CURRENT_BINARY_DIR}/test/round7.json.h
+    test/unitester.cpp)
```
style nit: Follow surrounding style:
```suggestion
    test/unitester.cpp
  )
```
?

hebasto reviewed (commented):
Approach ACK fabda022536c7a000a575bbb05059d27d29b5cca.

Related: https://github.com/bitcoin/bitcoin/pull/30901 improves handling of data files in build scripts.

hebasto reviewed (approved):
ACK faf23a8d0328f288d64dd697de4efedbc6970531.

l0rinc commented on src/univalue/test/unitester.cpp:
```diff
@@ -44,87 +96,64 @@ static void runtest(std::string filename, const std::string& jdata)
-        "round6.json",              // bare false
-        "round7.json",              // bare null
-};
+#define TEST_FILE(name) {#name, json_tests::name}
```
I understand why a macro is the smallest change to make `runtest` happy, but I think we can avoid that weird string parsing and introduce enums instead:
```C++
enum class TestType { Fail, Pass, RoundTrip };
```
and define the test as:
```C++
inline constexpr std::map<std::string_view, TestType> tests{
    {json_tests::fail1, TestType::Fail},
    {json_tests::fail2, TestType::Fail},
    {json_tests::fail3, TestType::Fail},
...
```
which could simplify `runtest` to:
```C++
static void runtest(std::string_view data, TestType type) {
    UniValue val;
    bool testResult = val.read(std::string{data});

    if (type == TestType::Fail) {
        assert(!testResult);
        return;
    }
    ...
```

sedited reviewed (approved):
Nice improvements, Re-ACK faedcc5fdabc508aaade9b461e44ea3a6670f36b

l0rinc reviewed (changes requested):
Thanks for the cleanup, please consider a few simplification and unification recommendations

hebasto reviewed (approved):
re-ACK faedcc5fdabc508aaade9b461e44ea3a6670f36b.

l0rinc committed:
build: Use character literals for generated headers to avoid narrowing

Use character literals instead of integer hex values (i.e. `'\x5b','\x0a', ...` instead of `0x5b, 0x0a, ...`) for generated headers.
This avoids C++11 narrowing warnings in a more concise way than using explicit char casts.

Extra whitespace is also removed between elements for brevity.

l0rinc commented on cmake/script/GenerateHeaderFromJson.cmake:
```diff
@@ -6,7 +6,7 @@ cmake_path(GET JSON_SOURCE_PATH STEM json_source_basename)
 file(READ ${JSON_SOURCE_PATH} hex_content HEX)
 string(REGEX REPLACE "................" "\\0\n" formatted_bytes "${hex_content}")
-string(REGEX REPLACE "[^\n][^\n]" "0x\\0, " formatted_bytes "${formatted_bytes}")
+string(REGEX REPLACE "[^\n][^\n]" "char(0x\\0)," formatted_bytes "${formatted_bytes}")
```
Extracted to https://github.com/bitcoin/bitcoin/pull/31547 - pease resolve the comment

l0rinc commented:
ACK faedcc5fdabc508aaade9b461e44ea3a6670f36b

Avoiding file reads during test setup is a step forward in portability 👍 
We can continue the refactors in follow-up PRs.

Commit:
test: clang-format -i src/univalue/test/unitester.cpp

l0rinc commented:
ACK faf7eac364fb7f421a649b483286ac8681d92b31

The changes since last ACK:  [simplify char casting](https://github.com/bitcoin/bitcoin/pull/31547) and reformat the affected sources.

fjahr commented:
tACK faf7eac364fb7f421a649b483286ac8681d92b31

Reviewed code and verified that tests are still running as intended.

sedited reviewed (approved):
Re-ACK faf7eac364fb7f421a649b483286ac8681d92b31

hebasto reviewed (approved):
Re-ACK faf7eac364fb7f421a649b483286ac8681d92b31. The commit, which modifies CMake scripts, has been replaced with the one from https://github.com/bitcoin/bitcoin/pull/31547, and a formatting commit has been added since my recent [review](https://github.com/bitcoin/bitcoin/pull/31542#pullrequestreview-2517189261).

achow101 commented:
ACK faf7eac364fb7f421a649b483286ac8681d92b31
